netbox/ipam/forms/model_forms.py

```diff
@@ -351,6 +351,18 @@ def clean(self):
                 'primary_for_parent', "Only IP addresses assigned to an interface can be designated as primary IPs."
             )
 
+        # Do not allow assigning a network ID or broadcast address to an interface.
+        if interface:
+            if address := self.cleaned_data.get('address'):
+                if address.ip == address.network and address.prefixlen not in (31, 32, 127, 128):
+                    self.add_error(
```

Comment:
Validation failures should raise `ValidationError` by convention. (I'm not sure why the check above is calling `add_error()`.